.github/workflows/release.yml

```diff
@@ -0,0 +1,49 @@
+name: Publish
+
+on:
+  # publish from the Releases page:
+  release:
+    types: [published]
+  # publish from the Actions page:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version (e.g. 2.0.3)'
+        required: true
+  # test publishing just for this PR
+  pull_request:
+
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v1
+    - name: Build tools
+      run: |
+          python -m pip install --upgrade pip
+          pip install build
+    - name: Replace dummy values in `setup.py` (only for `data-template`)
+      if: ${{ github.event.repository.name }} == "data-template"
+      run: perl -pi -e 's/\${dataset_name}/template/' setup.py
+    - name: Build
+      run: |
+          python -m build --wheel --sdist
+    - name: Publish to Github
+      uses: softprops/action-gh-release@v1
+      with:
+        files: 'dist/*'
+        fail_on_unmatched_files: true
+        tag_name: 2.0.3  #  ${{ github.event.inputs.version }} # in the workflow_dispatch case, make a new tag from the given input; in the published release case, this will be empty and will fall back to updating that release.
```

Comment:
Currently, in our `setup.py` we've set the option `use_scm_version=True`, which will use the most recent **git tag** as the _package_ version number.

There's a subtle bug here, though: For `workflow_dispatch`, we create a tag in the Publish step. But, this happens _after_ the build step, so the chosen tag won't be present for `python -m build` to pick up.

I think this means we have to try creating a tag prior to the build step?

Reply:
Specifically, this manifests in this PR as:

```
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         '0.1.dev1+g169fe48.d20220603' is an invalid value for Version. Error:  
         Can't use PEP 440 local versions. See                                  
        [ https://packaging.python.org/specifications/core](https://packaging.python.org/specifications/core-metadata)-metadata for m
         information.
```

The package is being build with `0.1.dev1`, when we want it to be built using the dummy example tag name I chose (`2.0.3`).

Reply:
Ahh! Fascinating! `use_scm_version=True` does _not_ do what I thought it did:

```
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         '2.0.4.dev0+g5929d76.d20220603' is an invalid value for Version. Error:
         Can't use PEP 440 local versions. See                                  
        [ https://packaging.python.org/specifications/core](https://packaging.python.org/specifications/core-metadata)-metadata for m
         information.
```

Setting the tag `2.0.3` prior to release will actually auto-generate a package with version number `2.0.4.dev0`!

My best guess is that `use_scm_version=True` is intended for the workflow described in the [README.md of the `pypi-publish` action](https://github.com/pypa/gh-action-pypi-publish#advanced-release-management):

> For example, you could implement a parallel workflow that pushes every commit to TestPyPI or your own index server, like devpi. For this, you'd need to (1) specify a custom repository_url value and (2) generate a unique version number for each upload so that they'd not create a conflict. The latter is possible if you use setuptools_scm package but you could also invent your own solution based on the distance to the latest tagged commit.

In other words, it's meant for rapidly generating unique version numbers during development (`dev0` -> `dev1` -> `dev2`), rather than generating the release number for stable releases.

This means that we should probably stop using `setuptools_scm` for building stable releases!

Reply:
Right! So, performing a find and replace on the `version="dev'` string gets us our desired result:

* https://test.pypi.org/project/spinalcordtoolbox-data-template/2.0.3/
* https://github.com/spinalcordtoolbox/data-template/releases/tag/2.0.3

Reply:
That doesn't make sense to me. I've used it lots of times and its always behaved. Take a look at https://github.com/ufoym/imbalanced-dataset-sampler/actions/runs/2371746877 to see for yourself.

If you want to do a test deploy you should use just tag it 'v1.0.3rc0' or 'v1.0.3rc1' -- that'll get tagged as a test build too -- particularly we should include this line

https://github.com/ufoym/imbalanced-dataset-sampler/actions/runs/2371746877

I'm on my phone atm. I'll need some time to look deeper about why setuptools-scm isn't working. But I really think we should use it, it really improves the workflow.

Reply:
> I'm on my phone atm. I'll need some time to look deeper about why setuptools-scm isn't working. But I really think we should use it, it really improves the workflow.


This seems to be intended behavior to me? From [`pypa/setuptools_scm`](https://github.com/pypa/setuptools_scm):

> Default versioning scheme
> 
> In the standard configuration setuptools_scm takes a look at three things:
> 
>     * latest tag (with a version number)
>     * the distance to this tag (e.g. number of revisions since latest tag)
>     * workdir state (e.g. uncommitted changes since latest tag)
> 
> and uses roughly the following logic to render the version:
> 
> * no distance and clean:
>     `{tag}`
> * distance and clean:
>     `{next_version}.dev{distance}+{scm letter}{revision hash}`
> * no distance and not clean:
>     `{tag}+dYYYYMMDD`
> * distance and not clean:
>     `{next_version}.dev{distance}+{scm letter}{revision hash}.dYYYYMMDD`
> 
> The next version is calculated by adding 1 to the last numeric component of the tag.

I presumed we were the "distance and clean" camp, hence:

* `{next_version}.dev{distance}+{scm letter}{revision hash}`, i.e.
* `2.0.4.dev0+g5929d76.d20220603`

Reply:
Why are we generating an extra commit? 

We should be able to use it like

* https://github.com/kousu/radicale-bsdauth/actions/runs/2203135595
* https://github.com/kousu/radicale-bsdauth/actions/runs/2203464045

In those I eventually decided to drop workflow_dispatch entirely: you just click the New Release button, github tags it, the build is triggered. If you want to do a changelog you just make sure its committed to master first.

Reply:
> But I really think we should use it, it really improves the workflow.

I don't understand the benefit of `setuptools_scm` when the person making the releases must explicitly specify the version number anyway. Could you explain your thought process, here?

Reply:
> Why are we generating an extra commit?

We aren't?

Reply:
...yeah right, I've read the whole thread now and the problem is that workflow_dispatch doesn't create the tag until the end and that complicates things. It's got a couple other difficult corner cases and it's just not worth trying to deal with imo!

Reply:
Ah, I see!! 

So, with `setuptools_scm`, if you go the `release` route, then you end up in the ["no distance and clean" case](https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889239510), at which point `{tag}` is used straight-up.

Reply:
I agree with you that we should drop `workflow_dispatch`. :)

Reply:
Ai yi yi. Glad we cleared that up. Sorry if I rushed to answer too fast!

Reply:
> Ah, I see!! I think I get why it's worked for you.
> 
> So, with `setuptools_scm`, if you go the `release` route, then you end up in the ["no distance and clean" case](https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889239510), at which point `{tag}` is used straight-up.
> 
> (In other words, they only append `.dev` if you're trying to release from an untagged commit.)

I thought we had cleared this up. But, I'm still running into the old `.dev0` automatic version bumping with `setuptools_scm`.

* I forked `data-template` to my personal GitHub ccount.
* I merged the PR as-is, and then re-added `setuptools_scm`.
* I then made a release, which [triggered the actions workflow](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true).
   * The commit is: https://github.com/joshuacwnewton/data-template/commit/e13b4f8673a3ddb338012191ee63cd016a791220
   * The version tag is: https://github.com/joshuacwnewton/data-template/tree/2022.6.6, which corresponds to the above commit.
   * In the workflow, we can confirm the correct commit is checked out [here](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true#step:3:108)
   * Yet, for some reason, instead of `2022.6.6`, the build command uses [`2022.6.7.dev0+ge13b4f8.d20220606`](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true#step:7:22)

I don't really know why my `setuptools_scm` setup is behaving differently compared to https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889243135. Is it because we're not using a `pyproject.toml` file? Is the version of `setuptools_scm` wrong? I'm going to try and investigate this more, but my first impressions of `setuptools_scm` have been that it's an complicated solution for the simple versioning use-case we're trying to go for here (as opposed to just `perl -pi -e`'ing the version number directly into `setup.py`). 

I'm still open to being convinced, but, grappling with this has been a little vexing. :P

Reply:
Oh wait. I think I might have figured out what the problem is!! `actions/checkout` checks out the **commit**, and not the **tag**. 

Hence the `.dev0`: We're on a git ref that's `0` distance away from the most recent tag, but we haven't actually checked out the tag itself.

Reply:
It's because of the `perl -pi`: you can't modify anything after the tag, that's no longer the tag.

I tried building off your tag locally and instead of running `perl -pi` I made the same edit, committed, and retagged it (pretending to be tomorrow), and that worked:

```
(a) p115628@joplin:~/src/data-template$ git log -p
commit 12aad1788069ba7ed277096cae3c9aad80524384 (HEAD, tag: 2022.6.7)
Author: Nick Guenther <nick.guenther@polymtl.ca>
Date:   Mon Jun 6 17:20:16 2022 -0400

    change dataset name

diff --git a/setup.py b/setup.py
index beebdc1..235e940 100644
--- a/setup.py
+++ b/setup.py
@@ -8,7 +8,7 @@ here = pathlib.Path(__file__).parent.resolve()
 import site; site.ENABLE_USER_SITE = True
 
 setup(
-  name='spinalcordtoolbox-data-${dataset_name}',
+  name='spinalcordtoolbox-data-t1',
   version="dev",
   description='Part of https://github.com/neuropoly/spinalcordtoolbox',
   long_description=(here / 'README.md').read_text(encoding='utf-8'),

commit e13b4f8673a3ddb338012191ee63cd016a791220 (tag: 2022.6.6)
Author: Joshua Newton <joshuacwnewton@gmail.com>
Date:   Mon Jun 6 16:29:53 2022 -0400

    `setup.py`: Re-add `setuptools_scm`
```

<details><summary>Successfully built spinalcordtoolbox-data-t1-2022.6.7.tar.gz and spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl</summary>

```
(a) p115628@joplin:~/src/data-template$ python -m build --wheel --sdist
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=40.8.0, setuptools_scm[toml], wheel)
* Getting dependencies for sdist...
* Installing packages in isolated environment... (setuptools, setuptools_scm[toml], wheel)
* Building sdist...
running sdist
running egg_info
writing src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO
writing dependency_links to src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt
writing top-level names to src/spinalcordtoolbox_data_t1.egg-info/top_level.txt
package init file 'src/spinalcordtoolbox/__init__.py' not found (or not a regular file)
package init file 'src/spinalcordtoolbox/data/__init__.py' not found (or not a regular file)
adding license file 'LICENSE.txt'
writing manifest file 'src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt'
running check
creating spinalcordtoolbox-data-t1-2022.6.7
creating spinalcordtoolbox-data-t1-2022.6.7/.github
creating spinalcordtoolbox-data-t1-2022.6.7/.github/workflows
creating spinalcordtoolbox-data-t1-2022.6.7/src
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data/dataset
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying files to spinalcordtoolbox-data-t1-2022.6.7...
copying .data-template-new.png -> spinalcordtoolbox-data-t1-2022.6.7
copying .gitignore -> spinalcordtoolbox-data-t1-2022.6.7
copying LICENSE.txt -> spinalcordtoolbox-data-t1-2022.6.7
copying README.md -> spinalcordtoolbox-data-t1-2022.6.7
copying pyproject.toml -> spinalcordtoolbox-data-t1-2022.6.7
copying setup.py -> spinalcordtoolbox-data-t1-2022.6.7
copying .github/workflows/release.yml -> spinalcordtoolbox-data-t1-2022.6.7/.github/workflows
copying src/spinalcordtoolbox/data/dataset/__init__.py -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data/dataset
copying src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/not-zip-safe -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/top_level.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
Writing spinalcordtoolbox-data-t1-2022.6.7/setup.cfg
Creating tar archive
removing 'spinalcordtoolbox-data-t1-2022.6.7' (and everything under it)
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=40.8.0, setuptools_scm[toml], wheel)
* Getting dependencies for wheel...
* Installing packages in isolated environment... (setuptools, setuptools_scm[toml], wheel)
* Building wheel...
running bdist_wheel
running build
running build_py<
package init file 'src/spinalcordtoolbox/__init__.py' not found (or not a regular file)
package init file 'src/spinalcordtoolbox/data/__init__.py' not found (or not a regular file)
running egg_info
writing src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO
writing dependency_links to src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt
writing top-level names to src/spinalcordtoolbox_data_t1.egg-info/top_level.txt
adding license file 'LICENSE.txt'
writing manifest file 'src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt'
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data/dataset
copying build/lib/spinalcordtoolbox/data/dataset/__init__.py -> build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data/dataset
running install_egg_info
Copying src/spinalcordtoolbox_data_t1.egg-info to build/bdist.linux-x86_64/wheel/spinalcordtoolbox_data_t1-2022.6.7-py3.10.egg-info
running install_scripts
adding license file "LICENSE.txt" (matched pattern "LICEN[CS]E*")
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox_data_t1-2022.6.7.dist-info/WHEEL
creating '/home/GRAMES.POLYMTL.CA/p115628/src/data-template/dist/tmphcd27lwe/spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'spinalcordtoolbox/data/dataset/__init__.py'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/LICENSE.txt'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/METADATA'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/WHEEL'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/top_level.txt'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Successfully built spinalcordtoolbox-data-t1-2022.6.7.tar.gz and spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl
```

</details>

Reply:
We just need to ask people need to manually fill in the name.

Maybe we can write some awesome actions script that edits setup.py to sync it with the repo's name on every commit. That would be cool, and in line with using git as the single source of truth, but maybe a bit Extra, y'know? I think doing step manually is good enough because how many times are we really going to be making a new dataset (vs making a new release of that dataset)?

Reply:
Oh of course. I missed the fact that editing `setup.py` means that the git repo is no longer in a clean slate. Thanks for that. :facepalm:

Reply:
> We just need to ask people need to manually fill in the name.

We already do! The `setup.py` edit step is only relevant for the `data-template` repo, and I only added it because otherwise the build on `data-template` would fail, hehe.

We can just straight-up remove the `perl` step, and allow the builds to fail on `data-template`? Since the problem goes away the moment people follow the README.

Reply:
From @kousu: 

> Ah good. I was pretty sure we already did...

Reply:
I tried out the template using PAM50, and all the steps went well: https://github.com/joshuacwnewton/data-PAM50/releases/tag/2022.6.6

I then tried installing the wheel and it behaved as expected.